File: scripts/etl_plastic_master.py

import pandas as pd
import json
import os
from supabase import create_client, Client
from dotenv import load_dotenv

load_dotenv('.env.local')
url = os.environ.get('NEXT_PUBLIC_SUPABASE_URL')
key = os.environ.get('SUPABASE_SERVICE_ROLE_KEY')
supabase = create_client(url, key)

df = pd.read_excel('./source_data/生産指示書/B. トレイデータ一覧表.xlsx', header=None, skiprows=4)

unique_plastics = {}
for idx, row in df.iterrows():
    zaisitsu = str(row.iloc[2]) if pd.notna(row.iloc[2]) else ''
    if zaisitsu == 'nan' or not zaisitsu or zaisitsu.strip() == '':
        continue
        
    thickness = str(row.iloc[3]) if pd.notna(row.iloc[3]) else ''
    width = str(row.iloc[4]) if pd.notna(row.iloc[4]) else ''
    taiden = str(row.iloc[5]) if pd.notna(row.iloc[5]) else ''
    silicon = str(row.iloc[6]) if pd.notna(row.iloc[6]) else ''
    tofu = str(row.iloc[7]) if pd.notna(row.iloc[7]) else ''
    
    code = f'{zaisitsu}_{thickness}_{width}_{taiden}_{silicon}_{tofu}'
    
    if code not in unique_plastics:
        family = zaisitsu.split('(')[0].split(' ')[0].strip()
        
        try:
            thick_val = float(thickness) if thickness.replace('.', '', 1).isdigit() else 0
        except: thick_val = 0
            
        try:
            width_val = int(float(width)) if width.replace('.', '', 1).isdigit() else 0
        except: width_val = 0
        
        unique_plastics[code] = {
            'plastic_code': code,
            'plastic_family': family,
            'plastic_subtype': zaisitsu,
            'thickness_mm': thick_val,
            'width_mm': width_val,
            'additive_text_raw': f'帯電:{taiden} ｼﾘｺﾝ:{silicon} 塗布:{tofu}',
            'is_active': True
        }

records = list(unique_plastics.values())
print(f'Total unique plastics to insert: {len(records)}')
families = list(set([r['plastic_family'] for r in records]))
print(f'Families: {families}')

try:
    res = supabase.table('plastic_master').upsert(records, on_conflict='plastic_code').execute()
    print(f'Successfully inserted {len(res.data)} records.')
except Exception as e:
    print(f'Error: {e}')

File: scripts/etl_plastic_master.mjs

import * as xlsx from 'xlsx';
import { createClient } from '@supabase/supabase-js';
import * as fs from 'fs';

const env = fs.readFileSync('.env.local', 'utf8');
const url = env.match(/NEXT_PUBLIC_SUPABASE_URL=([^\n]+)/)[1].trim();
const key = env.match(/SUPABASE_SERVICE_ROLE_KEY=([^\n]+)/)[1].trim();
const supabase = createClient(url, key);

const workbook = xlsx.readFile('./source_data/生産指示書/B. トレイデータ一覧表.xlsx');
const sheetName = workbook.SheetNames[0];
const sheet = workbook.Sheets[sheetName];
const data = xlsx.utils.sheet_to_json(sheet, { range: 3, defval: null });

const uniquePlastics = new Map();

for (const row of data) {
  const zaisitsu = row['材質'];
  if (!zaisitsu) continue;
  
  const thickness = row['厚み'];
  const width = row['ｼｰﾄ巾'];
  const taiden = row['帯電'];
  const silicon = row['ｼﾘｺﾝ'];
  const tofu = row['塗布'];
  
  // Create a unique key
  const code = \\-\-\-\-\-\\;
  
  if (!uniquePlastics.has(code)) {
    uniquePlastics.set(code, {
      plastic_code: code,
      plastic_family: zaisitsu.split('(')[0].trim(), // very naive split
      plastic_subtype: zaisitsu,
      thickness_mm: parseFloat(thickness) || 0,
      width_mm: parseInt(width) || 0,
      additive_text_raw: \帯電:\ ｼﾘｺﾝ:\ 塗布:\\,
      is_active: true
    });
  }
}

const records = Array.from(uniquePlastics.values());
console.log('Total unique plastics to insert:', records.length);
console.log('Sample families:', [...new Set(records.map(r => r.plastic_family))]);

async function run() {
  const { data: res, error } = await supabase.from('plastic_master').upsert(records, { onConflict: 'plastic_code' }).select();
  if (error) {
    console.error('Error inserting plastic_master:', error);
  } else {
    console.log('Successfully inserted', res.length, 'records.');
  }
}
run();

